src/security/Handshake.cc

```diff
@@ -274,39 +274,47 @@ Security::HandshakeParser::parseModernRecord()
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
     if (currentContentType != record.type) {
+        tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        fragments = record.fragment;
+        tkMessages.reset(record.fragment, true);
         currentContentType = record.type;
     } else {
-        fragments.append(record.fragment);
+        tkMessages.reset(tkMessages.leftovers().append(record.fragment), true);
     }
 
-    if (tkRecords.atEnd() && !done)
+    if (!done)
         parseMessages();
 }
 
 /// parses one or more "higher-level protocol" frames of currentContentType
 void
 Security::HandshakeParser::parseMessages()
 {
-    tkMessages.reset(fragments, false);
-    for (; !tkMessages.atEnd(); tkMessages.commit()) {
-        switch (currentContentType) {
-        case ContentType::ctChangeCipherSpec:
-            parseChangeCipherCpecMessage();
-            continue;
-        case ContentType::ctAlert:
-            parseAlertMessage();
-            continue;
-        case ContentType::ctHandshake:
-            parseHandshakeMessage();
-            continue;
-        case ContentType::ctApplicationData:
-            parseApplicationDataMessage();
-            continue;
+    try {
+        for (; !tkMessages.atEnd(); tkMessages.expectMore(false), tkMessages.commit()) {
+            switch (currentContentType) {
+                case ContentType::ctChangeCipherSpec:
+                    parseChangeCipherCpecMessage();
+                    continue;
+                case ContentType::ctAlert:
+                    parseAlertMessage();
+                    continue;
+                case ContentType::ctHandshake:
+                    parseHandshakeMessage();
+                    continue;
+                case ContentType::ctApplicationData:
+                    parseApplicationDataMessage();
+                    continue;
+            }
+            skipMessage("unknown ContentType msg [fragment]");
         }
-        skipMessage("unknown ContentType msg [fragment]");
+    }
+    catch (const Parser::BinaryTokenizer::InsufficientInput &) {
+        tkMessages.rollback();
+        Assure(tkMessages.expectMore());
+        debugs(83, 5, "need more data");
+        return;
```

Comment:
I assume that this "wait for more data" try/catch is not applicable to the first parseMessages() caller. Please either delete it completely (if possible) or move it to the second parseMessages() caller (otherwise).

Reply:
This try/catch was needed because BinaryTokenizer lacked an 'append()' method so we needed to rollback() before the makeshift reset(leftovers()+new_data). So I added append() and removed the try/catch.